README.md
# JED Checker

This extension is able to check your components, modules or plugins for common errors that will prevent you
from publishing your extension on the JED (Joomla! Extensions Directory).

If you are a developer and want to contribute to this extension you can fork this repo.

## Uploading your package

After installing this extension in your Joomla! backend, you can use it by uploading a Joomla! extension-package using
the upload-button. Once uploaded, the contents of the package (your files) will be checked against JED-rules.

## Adding rules

If you want to write a rule have a look a the `administrator/components/com_jedchecker/library/rules` folder.

You just need to add a new file with your rule, for example `yourrule.php`.

The file `yourrule.php` needs to have a class `jedcheckerRulesYourrule` and that class needs to have a
function that accepts the basedir as parameter. This is all - the component will automatically call
your rule check function.

If you are going to contribute your rule to the project, then make sure that it follows the joomla coding conventions
and that it passes the code sniffer: http://docs.joomla.org/Joomla_CodeSniffer

## Checking on existing files and folders

The extension also supports a scan of a pre-defined set of existing files and folders.
For this to work, add a list of folders to a textfile `tmp/jed_checker/local.txt`.
There should be a folder on each line.
Once the file exists, a "Check" button becomes visible in the jedchecker-toolbar. Just hit it.

Example `tmp/jed_checker/local.txt` file:

        components/com_jedchecker
        administrator/components/com_jedchecker
        plugins/system

## History of the Extension

This extension was previously maintained by Compojoom (Daniel Dimitrov). Other developers that collaborated with the original project were Denis Dulici (mijosoft.com), Riccardo Zorn (fasterjoomla.com), Bernard Toplak, and eaxs (projectfork.net).

Now, JED Checker is currently supported by Joomla (Open Source Matters).

## COPYRIGHT AND DISCLAIMER

Copyright (C) 2017 - 2021 Open Source Matters, Inc. All rights reserved.
Copyright (C) 2008 - 2016 compojoom.com . All rights reserved.

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program. If not, see http://www.gnu.org/licenses/.


CHANGELOG
<?php defined('_JEXEC') or die(); ?>

=====================================================================
JED Checker 3.0.0 - Released ??-???-2021
=====================================================================

^ Updated UI
+ Support of Joomla! 4.0
+ Rule XML Manifests
+ Rule XML Files references
+ Rule Language files
+ New inspections in INFO_XML rule
- Support all GPL-compatible and OSI-approved licenses in GPL rule
# Fixed false-positives in Framework rule
# Fixed false-positives in JAMSS rule
# Fixed false-positives in Encodings rule
- Other minor changes/improvements

=====================================================================
JED Checker 2.1.1 - 2.1.2 - Released 10-March-2019
=====================================================================

~ Copyright updated to Joomla! / OSM (source files, including the original authors and years)

=====================================================================
JED Checker 2.1.0 - Released 9-March-2019
=====================================================================
+ Compatibility with Joomla 4
+ Rule INFO_XML
+ Update Server Requirement
~ Copyright updated to Joomla! / OSM

=====================================================================
JED Checker 2.0 - Released 2-April-2017
=====================================================================
^ Removal of build tools
~ Copyright updated to Joomla! / OSM

=====================================================================
JED Checker 1.7 - Released 23-January-2017
=====================================================================
+ New Update Server rule
+ Addition of bootstrap collapses
- Removal of some deprecated Joomla Framework code
+ Add ACL
# Fixed some parameters and return types
~ Further bootstrapping
~ Language string updates
^ Minimum PHP version 5.3.10
^ error_reporting(0) now flags as error (was warning before)

=====================================================================
JED Checker 1.6 - Released 26-October-2015
=====================================================================
~ phpcs
# uploading zip files no longer displays a warning on Joomla 3.4.4
# updated link to JED listing requirements
+ added rule framework, jamss

=====================================================================
JED Checker 1.5 - Released 5-November-2015
=====================================================================
~ code cleaup and joomla's phpcs compatibility (at least for the rules)
# Fixed unnecessary GPL checks in php files with no code - thanks to Peter van Westen
# Fixed unnecessary JEXEC checks in php files with no code - thanks to Peter van Westen
+ added rule to check for error_reporting(0) - thanks to Denis Delici
# JFile not found

=====================================================================
LEGEND
=====================================================================
! Note
+ New feature or addition
^ Major change
~ Small change
$ Language change
* Security fix
# Bug fix
- Feature removal


administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
<?php
/**
 * @package    Joomla.JEDChecker
 *
 * @copyright  Copyright (C) 2017 - 2021 Open Source Matters, Inc. All rights reserved.
 * 			   Copyright (C) 2008 - 2016 compojoom.com . All rights reserved.
 * @author     Daniel Dimitrov <daniel@compojoom.com>
 *             eaxs <support@projectfork.net>
 *
 * @license    GNU General Public License version 2 or later; see LICENSE.txt
 */

defined('_JEXEC') or die('Restricted access');


// Include the rule base class
require_once JPATH_COMPONENT_ADMINISTRATOR . '/models/rule.php';


/**
 * class JedcheckerRulesXMLinfo
 *
 * This class searches all xml manifests for specific tags
 *
 * @since  1.0
 */
class JedcheckerRulesXMLinfo extends JEDcheckerRule
{
	/**
	 * The formal ID of this rule. For example: SE1.
	 *
	 * @var    string
	 */
	protected $id = 'INFO_XML';

	/**
	 * The title or caption of this rule.
	 *
	 * @var    string
	 */
	protected $title = 'COM_JEDCHECKER_INFO_XML';

	/**
	 * The description of this rule.
	 *
	 * @var    string
	 */
	protected $description = 'COM_JEDCHECKER_INFO_XML_DESC';

	/**
	 * Mapping of the plugin title prefix to the plugin group
	 *
	 * @var    string[]
	 */
	protected $pluginsGroupMap = array(
		'button' => 'editors-xtd',
		'editor' => 'editors',
		'smartsearch' => 'finder',
		'twofactorauthentication' => 'twofactorauth'
	);

	/**
	 * Initiates the search and check
	 *
	 * @return    void
	 */
	public function check()
	{
		// Find all XML files of the extension
		// @todo Load only main xml files and extra manifests from package
		$files = JFolder::files($this->basedir, '\.xml$', true, true);

		$manifestFound = false;

		// Iterate through all the xml files
		foreach ($files as $file)
		{
			// Try to find the license
			if ($this->find($file))
			{
				$manifestFound = true;
			}
		}

		if (!$manifestFound)
		{
			$this->report->addError('', JText::_('COM_JEDCHECKER_INFO_XML_NO_MANIFEST'));
		}
	}

	/**
	 * Reads a file and searches for the license
	 *
	 * @param   string  $file  - The path to the file
	 *
	 * @return boolean True if the manifest file was found, otherwise False.
	 */
	protected function find($file)
	{
		$xml = simplexml_load_file($file);

		// Failed to parse the xml file.
		// Assume that this is not a extension manifest
		if (!$xml)
		{
			return false;
		}

		// Check if this is an extension manifest
		// 1.5 uses 'install', 1.6+ uses 'extension'
		if ($xml->getName() !== 'extension')
		{
			return false;
		}

		// Get extension name (element)
		$type = (string) $xml['type'];

		if (isset($xml->element))
		{
			$extension = (string) $xml->element;
		}
		else
		{
			$extension = (string) $xml->name;

			if (isset($xml->files))
			{
				foreach ($xml->files->children() as $child)
				{
					if (isset($child[$type]))
					{
						$extension = (string) $child[$type];
					}
				}
			}
		}

		$extension = strtolower(JFilterInput::getInstance()->clean($extension, 'cmd'));

		if ($type === 'component' && strpos($extension, 'com_') !== 0)
		{
			$extension = 'com_' . $extension;
		}

		if ($type === 'plugin' && isset($xml['group']))
		{
			$extension = 'plg_' . $xml['group'] . '_' . $extension;
		}

		// Load the language of the extension (if any)
		$lang = JFactory::getLanguage();

		// Search for .sys.ini translation file
		$langDir = dirname($file);
		$langTag = 'en-GB';

		$lookupLangDirs = array();

		if (isset($xml->administration->files['folder']))
		{
			$lookupLangDirs[] = trim($xml->administration->files['folder'], '/') . '/language/' . $langTag;
		}

		if (isset($xml->files['folder']))
		{
			$lookupLangDirs[] = trim($xml->files['folder'], '/') . '/language/' . $langTag;
		}

		$lookupLangDirs[] = 'language/' . $langTag;

		if (isset($xml->administration->languages))
		{
			$folder = trim($xml->administration->languages['folder'], '/');

			foreach ($xml->administration->languages->language as $language)
			{
				if (trim($language['tag']) === $langTag)
				{
					$lookupLangDirs[] = trim($folder . '/' . dirname($language), '/');
				}
			}
		}

		if (isset($xml->languages))
		{
			$folder = trim($xml->languages['folder'], '/');

			foreach ($xml->languages->language as $language)
			{
				if (trim($language['tag']) === $langTag)
				{
					$lookupLangDirs[] = trim($folder . '/' . dirname($language), '/');
				}
			}
		}

		$lookupLangDirs[] = '';

		$lookupLangDirs = array_unique($lookupLangDirs);

		foreach ($lookupLangDirs as $dir)
		{
			$langSysFile = $langDir . '/' . ($dir === '' ? '' : $dir . '/') . $langTag . '.' . $extension . '.sys.ini';

			if (is_file($langSysFile))
			{
				$loadLanguage = new ReflectionMethod($lang, 'loadLanguage');
				$loadLanguage->setAccessible(true);
				$loadLanguage->invoke($lang, $langSysFile, $extension);
				break;
			}
		}

		// Get the real extension's name now that the language has been loaded
		$extensionName = $lang->_((string) $xml->name);

		$info[] = JText::sprintf('COM_JEDCHECKER_INFO_XML_NAME_XML', $extensionName);
		$info[] = JText::sprintf('COM_JEDCHECKER_INFO_XML_VERSION_XML', (string) $xml->version);
		$info[] = JText::sprintf('COM_JEDCHECKER_INFO_XML_CREATIONDATE_XML', (string) $xml->creationDate);

		$this->report->addInfo($file, implode('<br />', $info));

		// NM3 - Listing name contains “module” or “plugin”
		if (preg_match('/\b(?:module|plugin)\b/i', $extensionName))
		{
			$this->report->addError($file, JText::sprintf('COM_JEDCHECKER_INFO_XML_NAME_MODULE_PLUGIN', $extensionName));
		}

		if (stripos($extensionName, 'template') !== false)
		{
			$this->report->addWarning($file, JText::sprintf('COM_JEDCHECKER_INFO_XML_NAME_RESERVED_KEYWORDS', $extensionName));
		}

		// NM5 - Version in name/title
		if (preg_match('/(?:\bversion\b|\d\.\d)/i', $extensionName))
		{
			$this->report->addError($file, JText::sprintf('COM_JEDCHECKER_INFO_XML_NAME_VERSION', $extensionName));
		}

		if (stripos($extensionName, 'joomla') === 0)
		{
			$this->report->addError($file, JText::sprintf('COM_JEDCHECKER_INFO_XML_NAME_JOOMLA', $extensionName));
		}
		elseif (stripos($extensionName, 'joom') !== false)
		{
			$this->report->addWarning($file, JText::sprintf('COM_JEDCHECKER_INFO_XML_NAME_JOOMLA_DERIVATIVE', $extensionName));
		}

		$this->validateDomain($file, (string) $xml->authorUrl);

		if ($type === 'package' && (string) $xml->packagerurl !== (string) $xml->authorUrl)
		{
			$this->validateDomain($file, (string) $xml->packagerurl);
		}

		if ($type === 'component' && isset($xml->administration->menu))
		{
			$menuName = $lang->_(trim($xml->administration->menu));

			if ($extensionName !== $menuName)
			{
				$this->report->addWarning($file, JText::sprintf('COM_JEDCHECKER_INFO_XML_NAME_ADMIN_MENU', $menuName, $extensionName));
			}
		}

		if ($type === 'plugin')
		{
			$parts = explode(' - ', $extensionName, 2);
			$extensionNameGroup = isset($parts[1]) ? strtolower(preg_replace('/\s/', '', $parts[0])) : false;
			$group = (string) $xml['group'];

			if ($extensionNameGroup !== $group && $extensionNameGroup !== str_replace('-', '', $group)
				&& !(isset($this->pluginsGroupMap[$extensionNameGroup]) && $this->pluginsGroupMap[$extensionNameGroup] === $group)
			)
			{
				$this->report->addWarning($file, JText::sprintf('COM_JEDCHECKER_INFO_XML_NAME_PLUGIN_FORMAT', $extensionName));
			}
		}

		// All checks passed. Return true
		return true;
	}

	/**
	 * Check domain name against the list of approved domains
	 *
	 * @param   string $file Current file name
	 * @param   string $url  URL to validate
	 */
	protected function validateDomain($file, $url)
	{
		if (stripos($url, 'joom') !== false)
		{
			if (strpos($url, '//') !== false)
			{
				$domain = parse_url(trim($url), PHP_URL_HOST);
			}
			else
			{
				$pos = strpos($url, '/');
				$domain = ($pos === false) ? $url : substr($url, 0, $pos);
			}

			if (stripos($domain, 'joom') !== false)
			{
				// Remove "www." subdomain prefix
				$domain = preg_replace('/^www\./', '', $domain);

				// Approved domains from https://tm.joomla.org/approved-domains.html
				$approvedDomains = file(__DIR__ . '/xmlinfo/approved-domains.txt', FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);

				if (!in_array($domain, $approvedDomains, true))
				{
					$this->report->addError($file, JText::sprintf('COM_JEDCHECKER_INFO_XML_URL_JOOMLA_DERIVATIVE', $url));
				}
			}
		}
	}
}
